src/contract.rs

```diff
@@ -242,10 +248,103 @@ pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryM
         _ => authenticated_queries(deps, msg),
     }
 }
 
+fn query_balance_with_permit<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    signed: SignedPermit,
+    signature: Signature,
+) -> Result<Binary, StdError> {
+    if signed.msgs.len() != 1 {
+        return Err(StdError::generic_err(format!(
+            "Must sign exactly 1 permit message, got: {:?}.",
+            signed.msgs.len()
+        )));
+    }
+
+    if signed.msgs[0].r#type != "query_balance_permit" {
+        return Err(StdError::generic_err(format!(
+            "Type must be 'query_balance_permit', got: {:?}.",
+            signed.msgs[0].r#type
+        )));
+    }
+
+    let permit = &signed.msgs[0].value;
+
+    // Validate permit_user_id
+    let _permit_user_id = &permit.permit_user_id;
+    // TODO fail if permit_user_id is revoked
+
+    // Validate permit message
+    let query_balance_permit_msg = ReadonlyConfig::from_storage(&deps.storage)
+        .constants()?
+        .query_balance_permit_msg;
+    if permit.message != query_balance_permit_msg {
+        return Err(StdError::generic_err(format!(
+            "Message must be '{:?}', got: '{:?}'.",
+            query_balance_permit_msg, permit.message
+        )));
+    }
+
+    // Validate that account is derived from pubkey
+    let pubkey = signature.pub_key.value;
+    let account_from_pubkey = deps.api.human_address(&pubkey_to_account(&pubkey))?;
+
+    if account_from_pubkey != permit.query_balance_of {
+        return Err(StdError::generic_err(format!(
+            "Input account {:?} not the same as account {:?} that was derived from pubkey {:?}.",
+            permit.query_balance_of,
+            account_from_pubkey,
+            pubkey.clone()
+        )));
+    }
+
+    // Validate signature, source: https://github.com/enigmampc/SecretNetwork/blob/master/cosmwasm/packages/wasmi-runtime/src/crypto/secp256k1.rs#L49-L82
+    let signed_bytes = &to_binary(&signed)?.0;
+
+    let signed_bytes_hash = Sha256::digest(signed_bytes);
+    let msg = secp256k1::Message::from_slice(signed_bytes_hash.as_slice()).map_err(|err| {
+        StdError::generic_err(format!(
+            "Failed to create a secp256k1 message from signed_bytes: {:?}",
+            err
+        ))
+    })?;
+
+    let verifier = Secp256k1::verification_only();
+
+    // Create `secp256k1`'s types
+    let secp256k1_signature = secp256k1::Signature::from_compact(&signature.signature.0)
+        .map_err(|err| StdError::generic_err(format!("Malformed signature: {:?}", err)))?;
+    let secp256k1_pubkey = secp256k1::PublicKey::from_slice(pubkey.0.as_slice())
+        .map_err(|err| StdError::generic_err(format!("Malformed pubkey: {:?}", err)))?;
+
+    verifier
+        .verify(&msg, &secp256k1_signature, &secp256k1_pubkey)
+        .map_err(|err| {
+            StdError::generic_err(format!(
+                "Failed to verify signatures for the given permit: {:?}",
+                err
+            ))
+        })?;
```

Comment:
wait, is the code from 295-333 not a standardized method????

Reply:
Not in the context of a contract.

Reply:
See how a similar concept is implemented in DAI's ERC20 contract: https://github.com/makerdao/dss/blob/44330065999621834b08de1edf3b962f6bbd74c6/src/dai.sol#L118-L140

Reply:
_nod_ prolly worth making this more generic for arbitrary signed permits tbh

Reply:
Arbitrary to do what?